apps/spruce/src/utils/tasks/estimatedActivatedTasks.ts

```diff
@@ -0,0 +1,106 @@
+import { VariantTask } from "gql/generated/types";
+import { VariantTasksState } from "hooks/useConfigurePatch";
+import {
+  selectedStrings,
+  versionSelectedTasks,
+} from "hooks/useVersionTaskStatusSelect";
+
+type ValidInputTypes = Set<string> | versionSelectedTasks | VariantTasksState;
+
+export const getNumEstimatedActivatedTasks = (
+  inputType: ValidInputTypes,
+  map: { [key: string]: number },
+  variantsTasks?: Array<VariantTask>,
+): number => {
+  if (inputType instanceof Set) {
+    return countForSet(inputType, map);
+  }
+  if (isVariantTasksState(inputType)) {
+    return countForVariantTasksState(inputType, map, variantsTasks);
+  }
+  if (isVersionSelectedTasks(inputType)) {
+    return countForVersionSelectedTasks(inputType, map);
+  }
+  throw new Error("Unrecognized input type for counting generated tasks");
+};
+
+const countForSet = (
+  selectedTasks: Set<string>,
+  generatedTaskCounts: { [key: string]: number },
+): number => {
+  let numEstimatedActivatedGeneratedTasks = selectedTasks.size;
+  selectedTasks.forEach((key) => {
+    if (generatedTaskCounts[key]) {
+      numEstimatedActivatedGeneratedTasks += generatedTaskCounts[key];
+    }
+  });
+  return numEstimatedActivatedGeneratedTasks;
+};
+
+const countForVariantTasksState = (
+  selectedBuildVariantTasks: VariantTasksState,
+  generatedTaskCounts: { [p: string]: number },
+  variantsTasks?: Array<VariantTask>,
+): number => {
+  const existingTasks = new Set<string>();
+  let numEstimatedActivatedGeneratedTasks = 0;
+
+  variantsTasks?.forEach((variantTask) => {
+    const variant = variantTask.name;
+    variantTask.tasks.forEach((task) => {
+      existingTasks.add(`${variant}-${task}`);
+    });
+  });
+
+  Object.keys(selectedBuildVariantTasks).forEach((variant) => {
+    Object.keys(selectedBuildVariantTasks[variant]).forEach((task) => {
+      if (
+        selectedBuildVariantTasks[variant][task] &&
+        !existingTasks.has(`${variant}-${task}`)
+      ) {
+        numEstimatedActivatedGeneratedTasks += 1;
+        if (generatedTaskCounts[`${variant}-${task}`]) {
+          numEstimatedActivatedGeneratedTasks +=
+            generatedTaskCounts[`${variant}-${task}`];
+        }
+      }
+    });
+  });
+  return numEstimatedActivatedGeneratedTasks;
+};
```

Comment:
Same comment as above this can be written as (with a jsdoc comment)
```ts
const sumSelectedEstimatedTasks = (
  selectedBuildVariantTasks: VariantTasksState,
  variantsTasks: VariantTask[],
  generatedTaskCounts: GeneratedTaskCountResults[],
): number => {
  // Create a set of existing tasks from variantsTasks
  const existingTasks = new Set<string>(
    variantsTasks.flatMap((variantTask) =>
      variantTask.tasks.map((task) => `${variantTask.name}-${task}`),
    ),
  );

  // Sum the estimated tasks for selected tasks that are not in existingTasks
  return Object.entries(selectedBuildVariantTasks).reduce(
    (total, [variant, tasks]) =>
      Object.entries(tasks).reduce((innerTotal, [taskName, isSelected]) => {
        if (isSelected && !existingTasks.has(`${variant}-${taskName}`)) {
          const generatedTaskCount =
            generatedTaskCounts.find(
              (count) =>
                count.buildVariantName === variant &&
                count.taskName === taskName,
            )?.estimatedTasks || 0;
          return innerTotal + 1 + generatedTaskCount;
        }
        return innerTotal;
      }, total),
    0,
  );
};
```